File: benchmarks/server/driver/src/drivers/driver.ts
// The neutral surface every database implements. The harness never knows whether it is driving
// Sirannon or PostgreSQL. It asks a driver to apply a schema, seed rows, and run a read, a write,
// or a transaction, and the driver reaches its database through that database's own shipping
// client. SQL arrives with `?` placeholders; a driver renders the placeholder in its own style.
//
// A driver also reports what a transaction costs it in round trips, because that is a property of
// the client, not of the workload: one client sends the whole statement list in a single request,
// another walks BEGIN, each statement, and COMMIT over the socket. The harness discloses the number
// rather than letting a reader assume every engine pays the same postage for one operation.

import type { SeedTable } from '../workloads/workload.ts'

export type Dialect = 'sqlite' | 'postgres'

export interface TransactionStatement {
  sql: string
  params: unknown[]
}

export abstract class Driver {
  abstract readonly name: string
  abstract readonly delivery: string
  abstract readonly dialect: Dialect

  abstract connect(): Promise<void>
  abstract info(): Promise<Record<string, unknown>>
  abstract executeDdl(statements: string[]): Promise<void>
  abstract dropTables(tables: string[]): Promise<void>
  abstract seed(tables: SeedTable[]): Promise<void>
  abstract read(sql: string, params: unknown[]): Promise<void>
  abstract write(sql: string, params: unknown[]): Promise<void>
  abstract transaction(statements: TransactionStatement[]): Promise<void>
  abstract transactionRoundTrips(statementCount: number): number
  abstract close(): Promise<void>

  render(sql: string): string {
    return sql
  }

  insertSql(table: SeedTable): string {
    const placeholders = table.columns.map(() => '?').join(', ')
    const columns = table.columns.join(', ')
    return `INSERT INTO ${table.table} (${columns}) VALUES (${placeholders})`
  }
}

File: benchmarks/server/src/sirannon_bench/core/driver.py
"""The neutral surface every database implements.

The harness never knows whether it is driving Sirannon or PostgreSQL. It asks a driver to apply
a schema, seed rows, and run a read or a write, and the driver reaches its database through that
database's own shipping client. SQL arrives with ``?`` placeholders; a driver renders the
placeholder in its own style.
"""

from __future__ import annotations

from abc import ABC, abstractmethod

from .workloads import SeedTable


class Driver(ABC):
    name: str
    delivery: str
    dialect: str

    @abstractmethod
    async def connect(self) -> None:
        """Open the client and confirm the database is reachable."""

    @abstractmethod
    async def info(self) -> dict:
        """Return the engine version and the durability-relevant settings in force."""

    @abstractmethod
    async def execute_ddl(self, statements: list[str]) -> None:
        """Run each schema statement in order."""

    @abstractmethod
    async def drop_tables(self, tables: list[str]) -> None:
        """Drop each table if it exists, so a workload starts from a known-empty state."""

    @abstractmethod
    async def seed(self, tables: list[SeedTable]) -> None:
        """Bulk-load the seed rows for a workload."""

    @abstractmethod
    async def read(self, sql: str, params: list) -> None:
        """Run one read. The rows are discarded; the caller times the call."""

    @abstractmethod
    async def write(self, sql: str, params: list) -> None:
        """Run one write."""

    @abstractmethod
    async def close(self) -> None:
        """Release the client and its connections."""

    def render(self, sql: str) -> str:
        """Render ``?`` placeholders in the dialect's style. The default keeps ``?``."""
        return sql

    def insert_sql(self, table: SeedTable) -> str:
        placeholders = ", ".join("?" for _ in table.columns)
        columns = ", ".join(table.columns)
        return f"INSERT INTO {table.table} ({columns}) VALUES ({placeholders})"
